### PR TITLE
Build infrastructure - CMake and CI

### PR DESCRIPTION
**Pull request by vvolkl**
Added a MCParticle class just to test the technicalities - running the class generator, compiling the library, creating the dictionaries, etc. The config is pulled from HSF/tools, HEP-FCC/fcc-edm and AIDASoft/podio, and just a starting point.

It seems I still have trouble getting the CI to mount cvmfs, working on it.

### CODE REVIEW

**Comment by graeme-a-stewart** [comment]
Thanks a lot @vvolkl! Re. the CI, I actually always ended up giving up on the Travis images and rolling a docker container that had everything setup and working inside. It's a lot easier to debug that setup as one then has a lot more control over the machine.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+project(EDM4HEP VERSION 0.1.0)
+
+#--- Run the podio class generator and link library -------
+
+find_package(podio REQUIRED HINTS $ENV{PODIO})
+
+execute_process(COMMAND python ${podio_CMAKE_DIR}/../python/podio_class_generator.py edm4hep.yaml edm4hep edm4hep
+                WORKING_DIRECTORY ${CMAKE_SOURCE_DIR})
+
+find_library(PODIOLIB podio PATHS ${podio_DIR}/../lib)
```
Comment (drbenmorgan) [inline CMakeLists.txt]: The `find_library` shouldn't be needed after a `find_package`.  That probably indicates an issue with podio's "PodioConfig" file.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,111 @@
+cmake_minimum_required(VERSION 3.3)
+project(EDM4HEP VERSION 0.1.0)
+
+#--- Run the podio class generator and link library -------
+
+find_package(podio REQUIRED HINTS $ENV{PODIO})
+
+execute_process(COMMAND python ${podio_CMAKE_DIR}/../python/podio_class_generator.py edm4hep.yaml edm4hep edm4hep
```
Comment (drbenmorgan) [inline CMakeLists.txt]: Unless `find_package(podio ...)` sets up Python, there should probably be a `find_package(PythonInterp)` before this. Again, maybe Podio should be supplying a CMake function to generate the code from the YAML.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                WORKING_DIRECTORY ${CMAKE_SOURCE_DIR})
+
+find_library(PODIOLIB podio PATHS ${podio_DIR}/../lib)
+if (NOT PODIOLIB)
+    message(FATAL_ERROR "libpodio.so(dylib) cannot be found dynamically. Make sure you have set up your environment to use PODIO")
+endif()
+link_directories(${podio_LIBRARY_DIR})
```
Comment (drbenmorgan) [inline CMakeLists.txt]: Always better to use explicit `target_link_libraries` on the targets.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  ${podio_INCLUDE_DIRS}
+  )
+
+#--- Define basic build settings -----------------------------------------------
+# - Use GNU-style hierarchy for installing build products
+include(GNUInstallDirs)
+
+# Offer the user the choice of overriding the installation directories
+set(INSTALL_LIB_DIR lib CACHE PATH "Installation directory for libraries")
+set(INSTALL_BIN_DIR bin CACHE PATH "Installation directory for executables")
```
Comment (drbenmorgan) [inline CMakeLists.txt]: Both shouldn't be needed as GNUInstallDirs *should* be providing lib/bin install locations.

**Comment by graeme-a-stewart** [comment]
@drbenmorgan - PODIO lives here: <https://github.com/AIDASoft/podio>. Yep, the CMake is probably a bit long in the tooth now...

**Comment by drbenmorgan** [comment]
Thanks @graeme-a-stewart, I'll take a look (and I should have emphasised - don't take the above as a "review" just that's the easiest way to comment!)

**Comment by andresailer** [comment]
Someone needs to enable the travis service for this Repo

**Comment by vvolkl** [comment]
Thanks for the comments! @drbenmorgan those are all great points, we just merged this PR to get the CI to run, but I'll try to clean up the local things, and it's a good opportunity to improve things in PODIO also.

**Pull request merged**